### ISSUE TITLE
Support Tagging for VPC Endpoints (Gateway Endpoints, Interface Endpoints and Endpoint Services)

### ISSUE THREAD

**Issue by sshvetsov** [terraform-providers/terraform-provider-aws#8625]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS just announced support for tagging of VPC Endpoints (Gateway Endpoints, Interface Endpoints, and Endpoint Services).
* What's New: https://aws.amazon.com/about-aws/whats-new/2019/05/amazon-vpc-endpoints-now-support-tagging-for-gateway-endpoints-interface-endpoints-and-endpoint-services/

Tags can be used to search, filter, and organize VPC endpoints and endpoint services, but more importantly, tags can clearly organize the charges for these resources in the cost explorer and allow better cost attribution and chargebacks in environments with large shared accounts and multiple accounts.

### New or Affected Resource(s)

* aws_vpc_endpoint
* aws_vpc_endpoint_service

### Potential Terraform Configuration

```hcl
resource "aws_vpc_endpoint" "s3" {
  vpc_id            = "${aws_vpc.main.id}"
  service_name      = "com.amazonaws.us-west-2.s3"
  vpc_endpoint_type = "Gateway"

  tags = {
    Name        = "S3 VPC Endpoint Gateway"
    Environment = "production"
    CostCenter  = "abc"
    ManagedBy   = "terraform"
  }
}

resource "aws_vpc_endpoint" "app_interface" {
  vpc_id             = "${aws_vpc.main.id}"
  subnet_ids         = ["${local.vpce_subnet_ids}"]
  security_group_ids = ["${local.security_group_ids}"]
  service_name       = "${local.service_provider_name}"
  vpc_endpoint_type  = "Interface"

  tags = {
    Name        = "Custom Service VPC Endpoint Interface"
    Environment = "production"
    CostCenter  = "xyz"
    ManagedBy   = "terraform"
  }
}

resource "aws_vpc_endpoint_service" "app_service" {
  acceptance_required        = false
  network_load_balancer_arns = ["${aws_lb.test.arn}"]

  tags = {
    Name        = "Custom Service VPC Endpoint Service"
    Environment = "production"
    CostCenter  = "abc"
    ManagedBy   = "terraform"
  }
}

```

**Comment by ewbankkit** [comment]
We need this one so I'll take it on if nobody else has started.

**Comment by ewbankkit** [comment]
Also update the associated data sources.

**Comment by ewbankkit** [comment]
It looks like the tags on these resources are controlled via the EC2 [`DescribeTags`](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeTags.html), [`CreateTags`](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateTags.html) and [`DeleteTags`](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DeleteTags.html) APIs.

**Comment by ewbankkit** [comment]
Better support in the EC2 API now available with [AWS SDK v1.19.31](https://github.com/aws/aws-sdk-go/releases/tag/v1.19.31):
> Adding tagging support for VPC Endpoints and VPC Endpoint Services.

Requires:
* https://github.com/terraform-providers/terraform-provider-aws/pull/8649

**Comment by ewbankkit** [comment]
Affected resources:
* [`aws_codepipeline`](https://www.terraform.io/docs/providers/aws/r/codepipeline.html)
* [`aws_codepipeline_webhook`](https://www.terraform.io/docs/providers/aws/r/codepipeline_webhook.html)

**Comment by jakauppila** [comment]
Can this move forward now that the SDK update has been merged?

**Comment by pleochroic** [comment]
Hi, Any updates about it?

**Comment by nunofernandes** [comment]
Any ideas on how this is going? Any ETA?

**Comment by bflad** [comment]
Support for tag management of VPC Endpoints has been merged and will release with version 2.16.0 of the Terraform AWS Provider, likely tomorrow. 👍

**Comment by bflad** [comment]
This has been released in [version 2.16.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2160-june-20-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
